Add ESQuery wrapper for user data searches

esoergel opened the pull request:
<!--- Provide a link to the ticket or document which prompted this change -->

##### SUMMARY
This seems like it warrants a first-class abstraction, so I'm adding one in.  Being that this is [an analyzed field](https://github.com/dimagi/commcare-hq/blob/dffe10b0b4dc1ed29fed8b4f40184d0c7f631163/corehq/pillows/mappings/user_mapping.py#L212-L212), a `match` query is a better fit, as described [here](https://www.elastic.co/guide/en/elasticsearch/reference/2.4/query-dsl-term-query.html).

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### RISK ASSESSMENT / QA PLAN
<!-- Does this need QA before or after merge? Link QA ticket. -->

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

esoergel committed:
Add ESQuery wrapper for user data searches

proteusvacuum commented:
would be cool to add a test, but this looks useful!

esoergel committed:
Write a basic test for filtering by user data

esoergel commented:
Fair point.  Just took a crack at it.

esoergel commented on corehq/apps/es/users.py:
```diff
@@ -166,3 +167,13 @@ def role_id(role_id):
     return filters.term("is_active", active)
+
+
+def user_data(key, value):
```
Meant to ask - is this the best name for this filter?  Would it be helpful to call it `user_data_exact` or `filter_user_data` or something?

orangejenny replied:
Or, to throw in a wrench, what about `metadata`? HQ code doesn't access `user_data` directly, it's wrapped in the `metadata` property.

esoergel replied:
Oh interesting - I didn't know about that change.  I'm actually not convinced, though, since it's not wrapped in ES code, and they're not quite the same thing.  This query only operates on the raw data sent to ES, so the run-time transformations that happen to `metadata` won't actually be queryable.  The results returned from ES will also contain only `user_data`, so by switching to `metadata`, there'd be a gap between what's searched and what's returned.  Is the intent to pretend that `user_data` doesn't exist?

orangejenny replied:
The metadata thing is a recent change, part of [user data profiles](https://github.com/dimagi/commcare-hq/pull/28208)

From a human behavior perspective, I want there to be as few references to the identifier `user_data` as possible, because it should no longer be accessed directly in python. In a world where couch migrations and ES reindexes weren't such a pain, I'd have renamed it `_user_data` to indicate it shouldn't be used, and probably also would have then renamed the value in the ES index.

But from the perspective of just this section of code, it is dealing with something called `user_data`, so there's logic to naming it similarly.

I agree it's messy either way and defer to your judgment.

esoergel committed:
Make a note about the various field names

esoergel commented:
Just rebased some changes - sorry about that, things were getting messy and I figured it'd be easier to review like this.  Notably, I switched to calling this `metadata` on @orangejenny 's recommendation.  I tried setting up test users by passing a `user_data` kwarg to `create`, and it failed silently and perplexingly, convincing me of the veracity of Jenny's argument.

But good grief, getting an ES test set up properly in our system is arcane.

orangejenny commented:
> I tried setting up test users by passing a user_data kwarg to create, and it failed silently and perplexingly, convincing me of the veracity of Jenny's argument.

As you probably found, `create` now has a `metadata` argument, though it does still take in arbitrary `kwargs` and apply them to the user. Would it be helpful to have `create` throw an error if the caller passes in `user_data`?

esoergel commented:
> Would it be helpful to have create throw an error if the caller passes in user_data?

That seems reasonable.  Honestly, IMO `**kwargs` is inviting nothing but trouble.  Looks like anything it catches is blindly `setattr`'d onto the django user, which I can't imagine is something we're intentionally making use of.  (Such a change is out of scope for this PR, of course)

orangejenny commented:
@esoergel Cool, I'm happy to do that in another PR (not getting rid of kwargs, but throwing an error if it includes user_data).

esoergel commented:
@proteusvacuum would you mind re-reviewing with the new tests?

proteusvacuum commented on corehq/apps/es/tests/test_user_es.py:
```diff
@@ -0,0 +1,52 @@
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        initialize_index_and_mapping(get_es_new(), USER_INDEX_INFO)
+        cls.elasticsearch = get_es_new()
```
This probably works fine, but I think it would be better to use the same instance of `get_es_new()` in these 2 lines

```suggestion
        cls.elasticsearch = get_es_new()
        initialize_index_and_mapping(cls.elasticsearch, USER_INDEX_INFO)
```

proteusvacuum commented on corehq/apps/es/tests/test_user_es.py:
```diff
@@ -0,0 +1,52 @@
+            self._create_mobile_worker(metadata={'foo': 'bar'})
+            self._create_mobile_worker(metadata={'foo': 'baz'})
+            self._create_mobile_worker(metadata={'foo': 'womp', 'fu': 'bar'})
+        get_es_new().indices.refresh(USER_INDEX)
```
```suggestion
        self.elasticsearch.indices.refresh(USER_INDEX)
```

proteusvacuum commented on corehq/apps/es/tests/test_user_es.py:
```diff
@@ -0,0 +1,52 @@
+            self._create_mobile_worker(metadata={'foo': 'baz'})
+            self._create_mobile_worker(metadata={'foo': 'womp', 'fu': 'bar'})
+        get_es_new().indices.refresh(USER_INDEX)
+        self.assertEqual(UserES().metadata('foo', 'bar').count(), 1)
```
It would be ideal if you validated that the user returned is the one you expect, otherwise this test is just testing that _something_ is being returned.

proteusvacuum commented on corehq/apps/es/tests/test_user_es.py:
```diff
@@ -0,0 +1,52 @@
+            self._create_mobile_worker(metadata={'foo': 'baz'})
+            self._create_mobile_worker(metadata={'foo': 'womp', 'fu': 'bar'})
+        get_es_new().indices.refresh(USER_INDEX)
+        self.assertEqual(UserES().metadata('foo', 'bar').count(), 1)
```
It might be good to also validate that chaining this query works as expected

```py
self.assertItemsEqual(UserES().metadata('fu', 'bar').metadata('foo', 'womp').ids(), [...])
self.assertItemsEqual(UserES().metadata('foo', 'bar').metadata('fu', 'bar').ids(), [])
```

esoergel replied:
I wouldn't have thought of that - are you thinking because it's a nested query the filter chaining might be different than with any other type of filter?  Regardless, I added that in.

proteusvacuum replied:
Also as documentation for whether this AND's these queries together or OR's them.

esoergel committed:
reuse es instance

Co-authored-by: Farid Rener <proteusvacuum@users.noreply.github.com>

esoergel committed:
Flesh out tests a bit more
